cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java

```diff
@@ -70,13 +88,83 @@
   private final DataProcConf conf;
   private final ClusterControllerClient client;
   private final Compute compute;
+  private final String projectId;
 
   public static DataProcClient fromConf(DataProcConf conf) throws IOException, GeneralSecurityException {
-    ClusterControllerClient client = ClusterControllerClient.create(conf.getControllerSettings());
-    return new DataProcClient(conf, client, conf.getCompute());
+    String projectId = conf.getProjectId();
+    if (projectId == null) {
+      projectId = getProjectId();
+    }
+
+    String accountKey = conf.getAccountKey();
+    ClusterControllerClient client = getClusterControllerClient(accountKey);
+    Compute compute = getCompute(accountKey);
+
+    return new DataProcClient(projectId, conf, client, compute);
+  }
+
+  private static String getProjectId() throws IOException {
```

Comment:
will be good to add a comment for this method.

Reply:
done